warehouse/integrations/vulnerabilities/utils.py

```diff
@@ -0,0 +1,147 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from sqlalchemy import func
+from sqlalchemy.orm.exc import NoResultFound
+
+from warehouse.integrations import vulnerabilities
+from warehouse.integrations.vulnerabilities.models import VulnerabilityRecord
+from warehouse.packaging.models import Project, Release
+
+
+def _get_project(request, vuln_report: vulnerabilities.VulnerabilityReportRequest):
+    return (
+        request.db.query(Project)
+        .filter(
+            Project.normalized_name == func.normalize_pep426_name(vuln_report.project)
+        )
+        .one()
+    )
+
+
+def _get_release(request, project: Project, version):
+    return (
+        request.db.query(Release)
+        .filter(Release.project_id == project.id)
+        .filter(Release.canonical_version == version)
+        .one()
+    )
+
+
+def _get_vuln_record(
+    request, vuln_report: vulnerabilities.VulnerabilityReportRequest, origin
+):
+    return (
+        request.db.query(VulnerabilityRecord)
+        .filter(VulnerabilityRecord.id == vuln_report.vulnerability_id)
+        .filter(VulnerabilityRecord.source == origin)
+        .one()
+    )
+
+
+def _delete_vuln_record(request, vuln_record: VulnerabilityRecord):
+    request.db.delete(vuln_record)
+
+
+def _add_vuln_record(request, vuln_record: VulnerabilityRecord):
+    request.db.add(vuln_record)
+
+
+def _analyze_vulnerability(request, vulnerability_report, origin, metrics):
+    try:
+        report = vulnerabilities.VulnerabilityReportRequest.from_api_request(
+            request=vulnerability_report
+        )
+    except vulnerabilities.InvalidVulnerabilityReportRequest as exc:
+        metrics.increment(f"warehouse.vulnerabilities.{origin}.error.{exc.reason}")
+        raise
+
+    metrics.increment(f"warehouse.vulnerabilities.{origin}.valid")
+
+    try:
+        vulnerability_record = _get_vuln_record(request, report, origin)
+
+        if not report.versions:
+            # No versions indicates the vulnerability is no longer considered
+            # valid, so delete it.
+            _delete_vuln_record(request, vulnerability_record)
+            return
+
+    except NoResultFound:
+        if not report.versions:
+            return
+
+        vulnerability_record = VulnerabilityRecord(
+            id=report.vulnerability_id,
+            source=origin,
+            link=report.advisory_link,
+            aliases=report.aliases,
+        )
+        _add_vuln_record(request, vulnerability_record)
+
+    try:
+        project = _get_project(request, report)
+    except NoResultFound:
+        metrics.increment(f"warehouse.vulnerabilities.{origin}.error.project_not_found")
+        raise
+
+    for version in report.versions:
```

Comment:
I believe this may do an awful lot of SQL queries. Can we fetch all the releases at once ?

Reply:
Also, I believe if we have a project with 1000 releases (some projects issue a release at every commit) and a report with 1000 versions, the double loop we'll have will be quadratic (1000^2 iterations). This would be both quite readable and more efficient using sets:

```
report_versions = set(report.versions)
record_versions = {r.version for r in vulnerability_record.releases}

# Add missing releases
releases_to_add = request.db.query(Release).do_some_magic(
    versions in report_versions - record_versions
)
for release in releases_to_add:
    vulnerability_record.releases.append(release)

# Remove outdated releases
releases_to_remove = request.db.query(Release).do_some_magic(
    versions in record_versions - report_versions
)
for release in releases_to_remove
    vulnerability_record.releases.remove(release)
```

I'm not familiar enough with SQLAlchemy's API to tell if there is a way to do it in one go, but just to check, there's isn't a way to just reassign `vulnerability_record.releases` to a list of releases, right ?

Reply:
I want to second this concern. To my eye this rill create an "N+1" query situation. I believe the thing desired is something in https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html

Reply:
added a joined load for releases when fetching the top-level vulnerability record

Reply:
Thanks, it's still not super clear to me how efficient the whole situation will be, but this should help!